crates/repl/src/runtime_panel.rs

```diff
@@ -0,0 +1,344 @@
+}
+
+impl RuntimePanel {
+    pub fn load(
+        workspace: WeakView<Workspace>,
+        cx: AsyncWindowContext,
+    ) -> Task<Result<View<Self>>> {
+        cx.spawn(|mut cx| async move {
+            let view = workspace.update(&mut cx, |workspace, cx| {
+                cx.new_view::<Self>(|cx| {
+                    let focus_handle = cx.focus_handle();
+
+                    let fs = workspace.app_state().fs.clone();
+
+                    let runtime_manager = cx.new_model(|cx| RuntimeManager::new(fs.clone(), cx));
+                    RuntimeManager::set_global(runtime_manager.clone(), cx);
+
+                    let subscriptions = vec![
+                        cx.on_focus_in(&focus_handle, Self::focus_in),
+                        cx.on_focus_out(&focus_handle, Self::focus_out),
+                        cx.observe(&runtime_manager, Self::handle_update),
+                        cx.observe_global::<SettingsStore>(move |_this, cx| {
+                            let settings = JupyterSettings::get_global(cx);
+                            if settings.enabled && RuntimeManager::global(cx).is_none() {
+                                // todo!(): turn panel and operations on and off
+                            } else {
+                                // todo!()
+                            }
+                        }),
+                    ];
+
+                    Self {
+                        width: None,
+                        focus_handle,
+                        runtime_manager: runtime_manager.clone(),
+                        sessions: Default::default(),
+                        workspace: workspace.weak_handle(),
+                        _subscriptions: subscriptions,
+                    }
+                })
+            })?;
+
+            view.update(&mut cx, |this, cx| {
+                this.runtime_manager
+                    .update(cx, |runtime_manager, cx| runtime_manager.load(cx))
+            })?;
+
+            Ok(view)
+        })
+    }
+
+    pub fn handle_update(&mut self, _model: Model<RuntimeManager>, cx: &mut ViewContext<Self>) {
+        cx.notify();
+    }
+
+    fn focus_in(&mut self, cx: &mut ViewContext<Self>) {
+        cx.notify();
+    }
+
+    fn focus_out(&mut self, _event: FocusOutEvent, cx: &mut ViewContext<Self>) {
+        cx.notify();
+    }
+
+    // Gets the active selection in the editor or the current line
+    pub fn selection(&self, editor: View<Editor>, cx: &mut ViewContext<Self>) -> Range<Anchor> {
+        let editor = editor.read(cx);
+        let selection = editor.selections.newest::<usize>(cx);
+        let buffer = editor.buffer().read(cx).snapshot(cx);
+
+        let range = if selection.is_empty() {
+            let cursor = selection.head();
+
+            let line_start = buffer.offset_to_point(cursor).row;
+            let mut start_offset = buffer.point_to_offset(Point::new(line_start, 0));
+
+            // Iterate backwards to find the start of the line
```

Comment:
The whole manual math around looks quite suspicious, but maybe I'm unaware of the context this is run in but overall looks like a wheel reinvention — Zed is dealing with text coordinates, rows and columns already.

Inside every buffer, there's a Rope that is supposed to have rows and columns + Zed's editor is able to convert these coordinates into e.g. "display" rows and columns, after line wraps, inlay hints and other text transformations applied.

So, I feel that we should something less low-level here, but let's discuss the context first.

Reply:
Yeah maybe there's a better way to have at this. Basically I want to be able to run "repl:Run" to run the current line (no matter position) or any selection. Most of the setup here is tracking the range so that later runs that overlap will clear old ones.